### DISCUSSION TITLE
Bekomme die Firmware nicht geflasht / ans Laufen

### DISCUSSION THREAD

**Discussion by Sprudelverduenner** [dewenni/ESP_Buderus_KM271#39] [Q&A] [answered]
Hallo Gemeinde.

Ich habe ein neues Modul KM217 WiFi Version 0.6 erhalten.
Nun möchte ich gerne die Firmware von hier mit MQTT-SUpport ans Laufen bekommen.

Wenn ich mit dem ESPHome Flasher die andere Firmware km271-for-friends.bin auf den ESP32 flashe dann sehe ich im LOG wie der ESP bootet und den AccessPoint öffnet.

Wenn ich mit ESPHome die Firmware esp_buderus_km271_v3_0_0.bin flashe dann ist kein AccessPoint zu sehen und im LOG steht folgendes:

```
[21:41:06]rst:0x10 (RTCWDT_RTC_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
[21:41:06]configsip: 0, SPIWP:0xee
[21:41:06]clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
[21:41:06]mode:QIO, clock div:1
[21:41:06]load:0x3fff0018,len:4
[21:41:06]load:0x49c2be71,len:-2113451951
[21:41:06]1162 mmu set 00010000, pos 00010000
[21:41:06]1162 mmu set 00020000, pos 00020000
[21:41:06]1162 mmu set 00030000, pos 00030000
[21:41:06]1162 mmu set 00040000, pos 00040000
[21:41:06]1162 mmu set 00050000, pos 00050000
[21:41:06]1162 mmu set 00060000, pos 00060000
[21:41:06]1162 mmu set 00070000, pos 00070000
[21:41:06]1162 mmu set 00080000, pos 00080000
[21:41:06]1162 mmu set 00090000, pos 00090000
[21:41:06]1162 mmu set 000a0000, pos 000a0000
[21:41:06]1162 mmu set 000b0000, pos 000b0000
[21:41:06]1162 mmu set 000c0000, pos 000c0000
[21:41:06]1162 mmu set 000d0000, pos 000d0000
[21:41:06]1162 mmu set 000e0000, pos 000e0000
[21:41:06]1162 mmu set 000f0000, pos 000f0000
[21:41:06]1162 mmu set 00100000, pos 00100000
[21:41:06]ets Jun  8 2016 00:22:57
```

Ich weiss leider nicht mehr was ich nun tun soll, um die Firmware ans Laufen zu bekommen.

Gibt es einen Tipp?
Ich bin gerne bereit weitere Informationen zu geben.

Vielen Dank vorab.
Beste Grüße,
Sprudelverduenner

**Comment by dewenni** [comment]
Hi,
Bitte teste erstmal das neuste Release 3.2.1

Grüße Sven

**Comment by Sprudelverduenner** [comment]
sorry, habe mich verschrieben.

sowohl mit esp_buderus_km271_v3_0_0.bin als auch esp_buderus_km271_v3_2_1.bin das exakt gleiche Verhalten....

**Reply by dewenni** [reply]
Schau dir mal diese Diskussion an: https://github.com/dewenni/ESP_Buderus_KM271/discussions/32#discussioncomment-5942452

Vielleicht ist es bei dir das gleiche.

**Reply by dewenni** [reply]
Vielleicht muss ich die bin Datei auch noch anders erzeugen. Ich habe bisher einfach die firmware.bin verwendet, die beim Kompilieren von PlattformIO generiert wird. Für Updates funktioniert das ja auch offensichtlich gut.
Vielleicht benötigt man für eine Erstinstallation aber noch eine speziellere bin Datei mit Bootloader?

**Comment by Sprudelverduenner** [comment]
Guten Morgen,

genau diesen Ansatz verfolge ich seit 30 Minuten bei meiner Recherche....

Jetzt habe ich leider auch nicht die Größte Ahnung auf diesem Gebiet - bin halt Anwender und kein Programmierer.

Gibt es denn bei deiner PlatformIO eine partition.bin und eine bootloader.bin?
Wenn ich die hätte dann könnte ich mal versuchen die mit einzubinden.

Oder ich muss mich mit PlatformIO für den Flash beschäftigen - aber ich denke die Anderen hätten am Meisten etwas davon wenn wir den Flash mit einem einfachen Tool hinbekämen....

**Comment by dewenni** [comment]
In der Tat gibt es beim Kompilieren auch die anderen bin Dateien.  
Ich habe sie dir hier mal in einem Zip zusammen gepackt.

[Archive.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/12060811/Archive.zip)

Ich habe heute leider nicht die Zeit mich näher damit zu beschäftigen, aber ich werde das in den kommenden Tagen mal tun.

**Comment by Sprudelverduenner** [comment]
OK. Schonmal Danke vorab.

Dann werde ich auf Dich warten.
Mit den einzelnen Dateien der Flash war nicht erfolgreich.

PlatformIO hatte ich schonmal für ein anderes Projekt installiert gehabt - das ich für mich ein Buch mit 7 Siegeln, da steig ich nicht durch...
Beim compilieren kommt es da zur einer Fehlermeldung

**Reply by dewenni** [reply]
@Sprudelverduenner bitte teste mal diese Datei:
[merged-flash.bin.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/12061184/merged-flash.bin.zip)

**Reply by Sprudelverduenner** [reply]
leider negativ.....

Das sagt der LOG:

`Using 'COM3' as serial port.
Connecting................
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting.......
Detecting chip type... ESP32
Connecting...

Chip Info:
 - Chip Family: ESP32
 - Chip Model: ESP32-D0WD (revision 1)
 - Number of Cores: 2
 - Max CPU Frequency: 240MHz
 - Has Bluetooth: YES
 - Has Embedded Flash: NO
 - Has Factory-Calibrated ADC: YES
 - MAC Address: C4:DE:E2:A2:28:CC
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
 - Flash Size: 4MB
Unexpected error: The firmware binary is invalid (magic byte=FF, should be E9)`

**Comment by dewenni** [comment]
mit dem ESPHome Flash-Tool hatte ich das gleiche Problem.
Ich habe es dann mit dem esptool gemacht. Nutzt du Windows oder Mac?

**Reply by Sprudelverduenner** [reply]
Windows

**Reply by Sprudelverduenner** [reply, marked as answer]
Heureka... ich glaube ich habe es mit deinen Tipps geschafft.

Ich habe das ESP32 DOWNLOAD TOOL V3.8.5 genommen.

Und habe dann deine merge.bin ab 0x0 geflasht.

Nun blinkt die grüne LED und es zeigt sich ein Hotspot.

**Comment by Sprudelverduenner** [comment]
Evtl. bringt Dich dieser Link weiter ??
https://community.platformio.org/t/esp32-how-to-create-one-solid-firmware-package-that-will-include-bootloader-etc-in-a-single-file/24379

**Comment by dewenni** [comment]
So, ich habe ein wenig experimentiert und getestet.
Bisher habe ich immer nur die Applikation als bin-File (image) in den Releases angehangen.
Das eignet sich aber nicht für eine Initialinstallation sondern nur für ein OTA-Update.

Daher erzeuge ich nun auch zusätzlich immer noch ein bin-File bei dem ich bereits die `bootloader.bin`, `partitions.bin` und die `firmware.bin` zusammenfasse.  

Diese bin Datei kann dann mit einem ESP-Flash-Tool ab Adresse 0x0 geflashed werden.

Das ESPHome Flash-Tool macht hier aber leider Probleme, es gibt aber auch Alternativen die Funktionieren.
Ich habe die Beschreibung in der README.md entsprechend angepasst.

Es gibt also ab sofort immer zwei release Dateien:
- ESP_Buderus_KM271_vx.x.x_install.bin (für eine initial Installation)
- ESP_Buderus_KM271_vx.x.x_OTA_update.bin (für OTA Updates bei bestehender Applikation)

**Reply by Sprudelverduenner** [reply]
Vielen Lieben Dank für Deine Mühen.

Die MQTT-Daten sind auch schon im meinem FHEM gelandet.
Jetzt muss ich das Device in FHEM nur noch anpassen um auch von FHEM heraus Einstellungen vornehmen zu können.

**Reply by dewenni** [reply]
freut mich - drücke gerne noch den Star ⭐️ und watch 👁 um auf dem Laufenden zu bleiben.

Viel Erfolg und schönes Wochenende noch

**Comment by joeflower** [comment]
Ich hatte das gleich Problem. Mit PlatformIO gings nach ein paar Anpassungen (WLAN,COM-Port, OTA auskommentiert) sofort.

**Reply by dewenni** [reply]
mir war bisher nicht bewusst, dass man für die "Erstinstallation" auf einem ESP ein anderes Image braucht, als das was ich bisher veröffentlich habe.

Ich habe mein Release-Ablauf jetzt nochmal etwas verfeinert, so dass mir jetzt immer automatisch die beiden Images erzeugt werden.
Einmal für OTA Updates und einmal für das Flashen von einem ESP32 Board.

Damit sollte es nun einfacher sein ein neues Board ans laufen zu bekommen.
